### PR TITLE
update auction_end handling to reduce storage operations

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: modules/auction_manager/src/lib.rs [hunk 1]
```diff
@@ -734,134 +734,136 @@ impl<T: Trait> Module<T> {
...
+				if let Ok(amount) = T::CDPTreasury::swap_collateral_to_stable(
+					collateral_auction.currency_id,
+					collateral_auction.amount,
+					Zero::zero(),
+				) {
+					// swap successfully, will not deal
+					should_deal = false;
+
+					// refund stable currency to the last bidder, ignore result to continue
+					let _ = T::CDPTreasury::issue_debit(&bidder, bid_price, false)
```
Comment (xlc) [inline modules/auction_manager/src/lib.rs:766]:
```suggestion
					T::CDPTreasury::issue_debit(&bidder, bid_price, false)
```

File: modules/auction_manager/src/lib.rs [hunk 2]
```diff
@@ -734,134 +734,136 @@ impl<T: Trait> Module<T> {
...
+						.expect("debit pool cannot overflow; qed");
+
+					if collateral_auction.in_reverse_stage(amount) {
+						// refund extra stable currency to recipient, ignore result to continue
+						let refund_amount = amount
+							.checked_sub(collateral_auction.target)
+							.expect("ensured amount > target; qed");
+						let _ = T::CDPTreasury::issue_debit(&collateral_auction.refund_recipient, refund_amount, false)
```
Comment (xlc) [inline modules/auction_manager/src/lib.rs:774]:
```suggestion
						T::CDPTreasury::issue_debit(&collateral_auction.refund_recipient, refund_amount, false)
```

File: modules/auction_manager/src/lib.rs [hunk 3]
```diff
@@ -734,134 +734,136 @@ impl<T: Trait> Module<T> {
...
+	) {
+		if let Some((bidder, _)) = winner {
+			// decrease account ref of winner
+			system::Module::<T>::dec_ref(&bidder);
 
-				// deposit unbacked stable token to winner by CDP treasury, ignore Err
-				let _ = T::CDPTreasury::issue_debit(&bidder, surplus_auction.amount, false);
+			// issue native token to winner, ignore the result to continue
+			// TODO: transfer from RESERVED TREASURY instead of issuing
+			let _ = T::Currency::deposit(T::GetNativeCurrencyId::get(), &bidder, debit_auction.amount)
```
Comment (xlc) [inline modules/auction_manager/src/lib.rs:827]:
```suggestion
			T::Currency::deposit(T::GetNativeCurrencyId::get(), &bidder, debit_auction.amount)
```

File: modules/auction_manager/src/lib.rs [hunk 4]
```diff
@@ -734,134 +734,136 @@ impl<T: Trait> Module<T> {
...
+		auction_id: AuctionId,
+		surplus_auction: SurplusAuctionItem<T::BlockNumber>,
+		winner: Option<(T::AccountId, Balance)>,
+	) {
+		if let Some((bidder, bidder_price)) = winner {
+			// decrease account ref of winner
+			system::Module::<T>::dec_ref(&bidder);
+
+			// deposit unbacked stable token to winner by CDP treasury, ignore Err
+			let _ = T::CDPTreasury::issue_debit(&bidder, surplus_auction.amount, false)
```
Comment (xlc) [inline modules/auction_manager/src/lib.rs:853]:
```suggestion
			T::CDPTreasury::issue_debit(&bidder, surplus_auction.amount, false)
```

**Review by xlc — COMMENTED** [review]
I would like to see more solid. Saying it cannot overflow is not a proof.

File: modules/auction_manager/src/lib.rs [hunk 5]
```diff
@@ -734,134 +723,138 @@ impl<T: Trait> Module<T> {
...
+					collateral_auction.currency_id,
+					collateral_auction.amount,
+					Zero::zero(),
+				) {
+					// swap successfully, will not deal
+					should_deal = false;
+
+					// refund stable currency to the last bidder, ignore result to continue
+					T::CDPTreasury::issue_debit(&bidder, bid_price, false).expect(
+						"issue_debit calls currency::deposit which always resolves so there's no error here; qed",
```
Comment (xlc) [inline modules/auction_manager/src/lib.rs:756]: this is still not a solid proof. what happen if the implementation changed in future?
should just ignore the error, and add a comment say it currently cannot fail. and it is ok to ignore the error in this case.

Comment (zjb0807) [reply modules/auction_manager/src/lib.rs]: Bryan means this https://github.com/AcalaNetwork/Acala/issues/366

**Comment by ermalkaleci** [comment]
This PR addresses #356 
https://github.com/AcalaNetwork/Acala/pull/356/files#diff-92e2cb9814da46b1efd7bcd8b3868948R768
https://github.com/AcalaNetwork/Acala/pull/356/files#diff-92e2cb9814da46b1efd7bcd8b3868948R775
https://github.com/AcalaNetwork/Acala/pull/356/files#diff-92e2cb9814da46b1efd7bcd8b3868948R909
https://github.com/AcalaNetwork/Acala/pull/356/files#diff-92e2cb9814da46b1efd7bcd8b3868948R1009

File: modules/auction_manager/src/lib.rs [hunk 6]
```diff
@@ -1008,14 +999,16 @@ impl<T: Trait> AuctionManager<T::AccountId> for Module<T> {
 
 	fn cancel_auction(id: Self::AuctionId) -> DispatchResult {
 		if <CollateralAuctions<T>>::contains_key(id) {
```
Comment (zjb0807) [inline modules/auction_manager/src/lib.rs:1001]:
```
if <CollateralAuctions<T>>::contains_key(id) {
```
```suggestion
if let Some(collateral_auction) = <CollateralAuctions<T>>::take(id) {
```

**Review by zjb0807 — APPROVED** [review]

**Pull request merged**